tests/ui/lint/non_local_definitions.rs

```diff
@@ -0,0 +1,246 @@
+// check-pass
+// edition:2021
+
+#![feature(inline_const)]
+
+use std::fmt::{Debug, Display};
+
+struct Test;
+
+impl Debug for Test {
+    fn fmt(&self, _f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        todo!()
+    }
+}
+
+mod do_not_lint_mod {
+    pub trait Tait {}
+
+    impl super::Test {
+        fn hugo() {}
+    }
+
+    impl Tait for super::Test {}
+}
+
+const _: () = {
+    trait Uto {}
+
+    impl Uto for Test {} // the trait is local, don't lint
+};
+
+trait Uto1 {}
+enum Enum {
+    Discr = {
+        impl Uto1 for Test {}
+        //~^ WARN non local definition
+
+        1
+    }
+}
+
+trait Uto2 {}
+static A: u32 = {
+    impl Uto2 for Test {}
+    //~^ WARN non local definition
+
+    1
+};
+
+trait Uto3 {}
+const B: u32 = {
+    impl Uto3 for Test {}
+    //~^ WARN non local definition
+
+    #[macro_export]
+    macro_rules! m0 { () => { } };
+    //~^ WARN non local definition
+
+    trait Uto4 {}
+    impl Uto4 for Test {}
+
+    1
+};
+
+trait Uto5 {}
+fn main() {
+    #[macro_export]
+    macro_rules! m { () => { } };
+    //~^ WARN non local definition
+
+    impl Test {
+    //~^ WARN non local definition
+        fn foo() {}
+    }
+
+    let _array = [0i32; {
+        impl Test {
+        //~^ WARN non local definition
+            fn bar() {}
+        }
+
+        1
+    }];
```

Comment:
Could you also test this in the type? i.e.
```rust
// top level

type A = [u32; { ... }];
// and/or
fn a(_: [u32; { ... }]) -> [u32; { ... }] { ... }
```

Reply:
Done.